ddl/column_type_change_test.go

```diff
@@ -0,0 +1,303 @@
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a int unsigned, b int signed)")
+	tk.MustExec("insert into t(a, b) values (1, 1)")
+	tk.MustExec("alter table t modify column a int signed")
+	tk.MustExec("alter table t modify column b int unsigned")
+
+	// Modify column from small type to big type.
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a tinyint)")
+	tk.MustExec("alter table t modify column a smallint")
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a tinyint)")
+	tk.MustExec("insert into t(a) values (127)")
+	tk.MustExec("alter table t modify column a smallint")
+	tk.MustExec("alter table t modify column a mediumint")
+	tk.MustExec("alter table t modify column a int")
+	tk.MustExec("alter table t modify column a bigint")
+
+	// Modify column from big type to small type.
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a bigint)")
+	tk.MustExec("alter table t modify column a int")
+	tk.MustExec("alter table t modify column a mediumint")
+	tk.MustExec("alter table t modify column a smallint")
+	tk.MustExec("alter table t modify column a tinyint")
+
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (a bigint)")
+	tk.MustExec("insert into t(a) values (9223372036854775807)")
+	tk.MustGetErrCode("alter table t modify column a int", mysql.ErrDataOutOfRange)
+	tk.MustGetErrCode("alter table t modify column a mediumint", mysql.ErrDataOutOfRange)
+	tk.MustGetErrCode("alter table t modify column a smallint", mysql.ErrDataOutOfRange)
+	tk.MustGetErrCode("alter table t modify column a tinyint", mysql.ErrDataOutOfRange)
+}
+
+func (s *testColumnTypeChangeSuite) TestColumnTypeChangeStateBetweenInteger(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t (c1 int, c2 int)")
+	tk.MustExec("insert into t(c1, c2) values (1, 1)")
+	// Enable column change variable.
+	tk.Se.GetSessionVars().EnableChangeColumnType = true
+	defer func() {
+		tk.Se.GetSessionVars().EnableChangeColumnType = false
+	}()
+
+	// use new session to check meta in callback function.
+	internalTK := testkit.NewTestKit(c, s.store)
+	internalTK.MustExec("use test")
+
+	tbl := testGetTableByName(c, tk.Se, "test", "t")
+	c.Assert(tbl, NotNil)
+	c.Assert(len(tbl.Cols()), Equals, 2)
+	c.Assert(getModifyColumn(c, tk.Se.(sessionctx.Context), "test", "t", "c2"), NotNil)
+
+	originalHook := s.dom.DDL().GetHook()
+	defer s.dom.DDL().(ddl.DDLForTest).SetHook(originalHook)
+
+	hook := &ddl.TestDDLCallback{}
+	var checkErr error
+	hook.OnJobRunBeforeExported = func(job *model.Job) {
+		if tbl.Meta().ID != job.TableID {
+			return
+		}
+		switch job.SchemaState {
+		case model.StateNone:
+			tbl = testGetTableByName(c, internalTK.Se, "test", "t")
+			if tbl == nil {
+				checkErr = errors.New("tbl is nil")
+			}
+			if len(tbl.Cols()) != 2 {
```

Comment:
If `tbl == nil` is true, this case will panic.

Reply:
good catch